hooks/basic/collector.py

```diff
@@ -207,18 +207,76 @@ def on_context_changed(self, settings, item):
         """
 
         path = item.properties["path"]
-        is_sequence = item.properties["is_sequence"]
 
-        # get info for the item
-        item_info = self._get_item_info(settings, path, is_sequence)
-        work_path_template = item_info["work_path_template"]
+        item.properties["work_path_template"] = self._resolve_work_path_template(item.properties, path)
+
+        # Set the item's fields property
+        item.properties["fields"] = self._resolve_item_fields(item)
+
+
+    def _resolve_work_path_template(self, properties, path):
+        """
+        Resolve work_path_template from the properties.
+        The signature uses properties, so that it can resolve the template even if the item object hasn't been created.
+
+        :param properties: properties that have/will be used to build item object.
+        :param path: path to be used to get the templates, using template_from_path.
+        :return: Name of the template.
+        """
+
+        work_path_template = properties.get("work_path_template")
 
         # If defined, add the work_path_template to the item's properties
+        # set/evaluate the correct work_path_template
+        # First check if a work_path_template has been defined
         if work_path_template:
-            item.properties["work_path_template"] = work_path_template
+            work_tmpl = self.parent.get_template_by_name(work_path_template)
+            if not work_tmpl:
+                # this template was not found in the template config!
+                raise TankError("The template '%s' does not exist!" % work_path_template)
 
-        # Set the item's fields property
-        item.properties["fields"] = self._resolve_item_fields(item)
+            # update the field with correct value so that we can use it everytime for this item
+            work_path_template = work_tmpl.name
+
+        # Else see if the path matches an existing template
+        else:
+            # let's try to check if this path fits into any known template
+            work_tmpl = self.sgtk.template_from_path(path)
+            if not work_tmpl:
+                # this path doesn't map to any known templates!
+                self.logger.warning("Cannot find a matching template for path: %s" % path)
```

Comment:
Error reporting should use the item name (aka display_name)

Reply:
Item might not exist for since this method is also used when I'm running `_get_item_context_from_path`. Should I handle template resolution separately in the context method?

Reply:
You can also get the item name as follows: `display_name = publisher.util.get_publish_name(path)`